src/server/game/AI/PlayerAI/PlayerAI.cpp

```diff
@@ -705,6 +705,29 @@ void PlayerAI::DoAutoAttackIfReady()
         DoMeleeAttackIfReady();
 }
 
+void PlayerAI::CancelAllShapeshifts()
+{
+    std::list<AuraEffect*> const& shapeshiftAuras = me->GetAuraEffectsByType(SPELL_AURA_MOD_SHAPESHIFT);
+    std::set<Aura*> removableShapeshifts;
+    for (AuraEffect* auraEff : shapeshiftAuras)
+    {
+        Aura* aura = auraEff->GetBase();
+        if (!aura)
+            continue;
+        SpellInfo const* auraInfo = aura->GetSpellInfo();
+        if (!auraInfo)
+            continue;
+        if (auraInfo->HasAttribute(SPELL_ATTR0_CANT_CANCEL))
```

Comment:
Only that line is present there because it's called on spell effect execution. If the spell executes, it's already been checked against all auras the player owns and isn't prevented by any existing (involuntary) shapeshifts.

This code should stay for consistency with future revisions, though I agree that it's superfluous on 3.3.5a.